File: server/lib/kb_io.js
/**
 * KB 写入安全包装（Node 版）：每次落盘前先把当前磁盘上的版本快照一份。
 * 与 server/lib/kb_io.py 行为完全一致，备份目录共享。
 *
 * 用法：
 *   const { saveKbSafely } = require('./lib/kb_io');
 *   saveKbSafely(kb, '/abs/path/to/kb.json');
 */
const fs = require('fs');
const path = require('path');

function ts() {
  const d = new Date();
  const pad = (n, w = 2) => String(n).padStart(w, '0');
  return `${d.getUTCFullYear()}${pad(d.getUTCMonth() + 1)}${pad(d.getUTCDate())}-` +
         `${pad(d.getUTCHours())}${pad(d.getUTCMinutes())}${pad(d.getUTCSeconds())}-` +
         `${pad(d.getUTCMilliseconds(), 3)}000`;  // 6 位仿 microseconds
}

function backupsDir(targetPath) {
  const abs = path.resolve(targetPath);
  const baseDir = path.dirname(abs);
  const stem = path.basename(abs, path.extname(abs));
  return path.join(baseDir, '.backups', stem);
}

function snapshot(targetPath) {
  const abs = path.resolve(targetPath);
  if (!fs.existsSync(abs)) return null;
  const bdir = backupsDir(abs);
  fs.mkdirSync(bdir, { recursive: true });
  const bpath = path.join(bdir, `${ts()}.json`);
  fs.copyFileSync(abs, bpath);
  return bpath;
}

function pruneBackups(targetPath, maxKeep = 20) {
  const bdir = backupsDir(targetPath);
  if (!fs.existsSync(bdir)) return;
  const files = fs.readdirSync(bdir)
    .filter(f => f.endsWith('.json'))
    .map(f => ({ p: path.join(bdir, f), m: fs.statSync(path.join(bdir, f)).mtimeMs }))
    .sort((a, b) => a.m - b.m);
  const excess = files.length - maxKeep;
  if (excess > 0) {
    for (let i = 0; i < excess; i++) {
      try { fs.unlinkSync(files[i].p); } catch { /* swallow */ }
    }
  }
}

/**
 * 备份现有 → 写新内容（先 .tmp 再 rename）→ 修剪旧备份。
 * 返回 { backup, target }。
 */
function saveKbSafely(kb, targetPath, maxBackups = 20) {
  const abs = path.resolve(targetPath);
  const backup = snapshot(abs);
  const tmp = abs + '.tmp';
  fs.writeFileSync(tmp, JSON.stringify(kb, null, 2), 'utf8');
  fs.renameSync(tmp, abs);
  pruneBackups(abs, maxBackups);
  return { backup, target: abs };
}

function listBackups(targetPath) {
  const bdir = backupsDir(targetPath);
  if (!fs.existsSync(bdir)) return [];
  return fs.readdirSync(bdir)
    .filter(f => f.endsWith('.json'))
    .map(f => ({ p: path.join(bdir, f), m: fs.statSync(path.join(bdir, f)).mtimeMs }))
    .sort((a, b) => b.m - a.m)
    .map(o => o.p);
}

module.exports = { saveKbSafely, snapshot, pruneBackups, listBackups, backupsDir };

File: server/lib/kb_io.py
"""
KB 写入安全包装：每次落盘前先把当前磁盘上的版本快照一份。

为什么要这玩意：
  scene KB 没有 git，多个脚本都会 load → mutate → save，一个脚本读到旧状态再
  覆盖回去就会悄悄抹掉别人的工作（这就是 24 个手写富化曾经一夜消失的原因）。
  每次写之前都备份，至少给"恢复到 5 分钟前"留了一条退路。

约定：
  - 备份目录: <kb_dir>/.backups/<filename_no_ext>/
  - 文件名:   <ts>.json     （ts = 当前 UTC YYYYMMDD-HHMMSS-<microseconds>）
  - 保留数:   默认 20，超出按时间最早的删
  - 写入是原子的：先写到 .tmp，rename 到目标

用法：
    from lib.kb_io import save_kb_safely
    save_kb_safely(kb, r'C:\\...\\house_of_dragon_05.json')
"""
import json
import os
import shutil
import datetime


def _ts():
    return datetime.datetime.now(datetime.timezone.utc).strftime('%Y%m%d-%H%M%S-%f')


def _backups_dir(target_path):
    target_path = os.path.abspath(target_path)
    base_dir = os.path.dirname(target_path)
    stem = os.path.splitext(os.path.basename(target_path))[0]
    return os.path.join(base_dir, '.backups', stem)


def snapshot(target_path):
    """把当前磁盘上的 target_path 复制一份到备份目录。返回备份路径或 None（文件不存在时）。"""
    target_path = os.path.abspath(target_path)
    if not os.path.exists(target_path):
        return None
    bdir = _backups_dir(target_path)
    os.makedirs(bdir, exist_ok=True)
    bpath = os.path.join(bdir, f'{_ts()}.json')
    shutil.copy2(target_path, bpath)
    return bpath


def prune_backups(target_path, max_keep=20):
    """只保留最新 max_keep 个备份；多的从最早删。"""
    bdir = _backups_dir(target_path)
    if not os.path.isdir(bdir):
        return
    files = sorted(
        (os.path.join(bdir, f) for f in os.listdir(bdir) if f.endswith('.json')),
        key=os.path.getmtime,
    )
    excess = len(files) - max_keep
    if excess > 0:
        for f in files[:excess]:
            try:
                os.remove(f)
            except OSError:
                pass


def save_kb_safely(kb, target_path, max_backups=20):
    """
    备份现有 → 写新内容（原子 rename）→ 修剪旧备份。
    返回 (backup_path_or_None, target_path)。
    """
    bpath = snapshot(target_path)
    target_path = os.path.abspath(target_path)
    tmp_path = target_path + '.tmp'
    with open(tmp_path, 'w', encoding='utf-8') as f:
        json.dump(kb, f, ensure_ascii=False, indent=2)
    os.replace(tmp_path, target_path)  # 原子替换
    prune_backups(target_path, max_keep=max_backups)
    return bpath, target_path


def list_backups(target_path):
    """返回该 KB 的所有备份路径，按时间从新到旧。"""
    bdir = _backups_dir(target_path)
    if not os.path.isdir(bdir):
        return []
    files = sorted(
        (os.path.join(bdir, f) for f in os.listdir(bdir) if f.endswith('.json')),
        key=os.path.getmtime,
        reverse=True,
    )
    return files
